File: solutions/cpp/Lowest_common_ancestor_no_parent.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Binary_tree_prototype.h"

using std::cout;
using std::endl;
using std::pair;
using std::unique_ptr;

pair<int, BinaryTreeNode<int>*> LCAHelper(
    const unique_ptr<BinaryTreeNode<int>>&,
    const unique_ptr<BinaryTreeNode<int>>&,
    const unique_ptr<BinaryTreeNode<int>>&);

// @include
BinaryTreeNode<int>* LCA(const unique_ptr<BinaryTreeNode<int>>& tree,
                         const unique_ptr<BinaryTreeNode<int>>& node0,
                         const unique_ptr<BinaryTreeNode<int>>& node1) {
  return LCAHelper(tree, node0, node1).second;
}

// Returns a pair of int and node pointer; int field is 0, 1, or 2 depending
// on how many of node0 and node1 are present in tree. If both are present in
// tree, the node pointer is a common ancestor. It may not be the LCA
// initially, but it will be LCA when the algorithm terminates.
pair<int, BinaryTreeNode<int>*> LCAHelper(
    const unique_ptr<BinaryTreeNode<int>>& tree,
    const unique_ptr<BinaryTreeNode<int>>& node0,
    const unique_ptr<BinaryTreeNode<int>>& node1) {
  if (tree == nullptr) {
    return {0, nullptr};
  }

  auto left_result = LCAHelper(tree->left, node0, node1);
  if (left_result.first == 2) {  // Found both nodes in the left subtree.
    return left_result;
  }
  auto right_result = LCAHelper(tree->right, node0, node1);
  if (right_result.first == 2) {  // Found both nodes in the right subtree.
    return right_result;
  }
  int num_target_nodes =
      left_result.first + right_result.first + (tree == node0 || tree == node1);
  return {num_target_nodes, num_target_nodes == 2 ? tree.get() : nullptr};
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> tree = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{3, nullptr, nullptr});
  tree->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{2, nullptr, nullptr});
  tree->left->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{1, nullptr, nullptr});
  tree->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{5, nullptr, nullptr});
  tree->right->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{4, nullptr, nullptr});
  tree->right->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{6, nullptr, nullptr});
  // should output 3
  auto* x = LCA(tree, tree->left, tree->right);
  assert(x->data == 3);
  cout << x->data << endl;
  // should output 5
  x = LCA(tree, tree->right->left, tree->right->right);
  assert(x->data == 5);
  cout << x->data << endl;
  // should output 5
  x = LCA(tree, tree->right, tree->right->right);
  assert(x->data == 5);
  cout << x->data << endl;
  // should output 3
  x = LCA(tree, tree->left->left, tree->right->right);
  assert(x->data == 3);
  cout << x->data << endl;
  // should output 3
  x = LCA(tree, tree->left->left, tree);
  assert(x->data == 3);
  cout << x->data << endl;
  return 0;
}

File: solutions/python/lowest_common_ancestor_no_parent.py
from binary_tree_prototype import BinaryTreeNode
import collections


# @include
def lca(tree, node0, node1):
    Status = collections.namedtuple('Status', ('num_target_nodes', 'ancestor'))

    # Returns an object consisting of an int and a node. The int field is 0,
    # 1, or 2 depending on how many of {node0, node1} are present in tree. If
    # both are present in tree, when ancestor is assigned to a non-null value,
    # it is the LCA.
    def lca_helper(tree, node0, node1):
        if not tree:
            return Status(0, None)

        left_result = lca_helper(tree.left, node0, node1)
        if left_result.num_target_nodes == 2:
            # Found both nodes in the left subtree.
            return left_result
        right_result = lca_helper(tree.right, node0, node1)
        if right_result.num_target_nodes == 2:
            # Found both nodes in the right subtree.
            return right_result
        num_target_nodes = (
            left_result.num_target_nodes + right_result.num_target_nodes + int(
                tree in (node0, node1)))
        return Status(num_target_nodes, tree if num_target_nodes == 2 else None)

    return lca_helper(tree, node0, node1).ancestor
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode(3)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(1)
    tree.right = BinaryTreeNode(5)
    tree.right.left = BinaryTreeNode(4)
    tree.right.right = BinaryTreeNode(6)
    # should output 3
    x = lca(tree, tree.left, tree.right)
    assert x.data == 3
    print(x.data)
    # should output 5
    x = lca(tree, tree.right.left, tree.right.right)
    assert x.data == 5
    print(x.data)
    # should output 5
    x = lca(tree, tree.right, tree.right.right)
    assert x.data == 5
    print(x.data)
    # should output 3
    x = lca(tree, tree, tree.left.left)
    assert x.data == 3
    print(x.data)
    # should output 2
    x = lca(tree, tree.left, tree.left.left)
    assert x.data == 2
    print(x.data)


if __name__ == '__main__':
    main()
